File: python/linkedlist/LinkedList.py
class Node:

    def __init__(self, e, next=None):
        self.e = e
        self.next = next

    def __str__(self):
        return str(self.e)


class LinkedList:

    def __init__(self):

        self.dummy_head = Node(None, None)
        self.size = 0

    def get_size(self) -> int:
        return self.size

    def is_empty(self) -> bool:
        return self.size == 0

    def add_first(self, e):
        self.add(0, e)

    def add_last(self, e):
        self.add(self.size, e)

    def add(self, index: int,  e):
        assert 0 <= index <= self.size  , "Add Failed. Illegal index"

        prev = self.dummy_head
        for _ in range(index):
            prev = prev.next

        # node = Node(e)
        # node.next = prev.next
        # prev.next = node

        prev.next = Node(e, prev.next)
        self.size += 1

    def get(self, index: int):
        prev = self.dummy_head.next
        for _ in range(index):
            prev = prev.next

        return prev.e

    def get_first(self):
        return self.get(0)

    def get_last(self):
        return self.get(self.size - 1)

    def set(self, index: int , e):
        assert 0 <= index <= self.size  , "Add Failed. Illegal index"

        prev = self.dummy_head.next
        for _ in range(index):
            prev = prev.next
        prev.e = e

    def contains(self, e) -> bool:
        cur = self.dummy_head.next

        while cur:
            if cur.e == e:
                return True
            cur = cur.next
        return False

    def delete(self, index: int):
        assert 0 <= index < self.size, "Add Failed. Illegal index"

        prev = self.dummy_head
        for _ in range(index):
            prev = prev.next

        del_node = prev.next
        prev.next = del_node.next
        del_node.next = None
        self.size -= 1
        return del_node

    def remove_first(self):
        return self.delete(0)

    def remove_last(self):
        return self.delete(self.size-1)

    def __str__(self):
        res = '[dummy'
        cur = self.dummy_head.next

        while cur:
            res += " -> "
            res += str(cur.e)
            cur = cur.next

        res += ' ->NULL ]'
        return res


                             
if __name__ == '__main__':
    print("Trie------------")

    ll = LinkedList()
    for i in range(10):
        ll.add_first(i)

    ll.add(1, 11)

    node = ll.dummy_head
    while node:
        print(node)
        node = node.next


    print("get({}) {} ".format(4 , ll.get(4)))

    ll.set(4, 99)
    print("get({}) {} ".format(4 , ll.get(4)))

    print(" get first ", ll.get_first())
    print(" get last  ", ll.get_last())


    print('contains: ', ll.contains(99))

    print(ll)
    ll.remove_first()
    ll.remove_last()
    print(ll)

File: src/com/dikers/linked/LinkedList.java
package com.dikers.linked;

public class LinkedList<E> {
    private class Node{
        public Node next;
        public E e;

        public Node(E e, Node next){
            this.e = e;
            this.next = next;
        }
        public Node(E e){
            this(e, null);
        }

        public Node(){
            this(null, null);
        }


        @Override
        public String toString() {
            return "N{" +
                    "e=" + e +
                    '}';
        }
    }


    private int size;
    private Node dummyHead;

    public LinkedList(){
        dummyHead = new Node(null, null);
        size = 0;
    }
    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public void addFirst(E e){
//        Node node = new Node(e);
//        node.next = head;
//        head = node;
        add(0,e);

    }
    // 在链表的index(0-based)位置添加新的元素e
    // 在链表中不是一个常用的操作，练习用：）
//    public void add(int index, E e){
//
//        if(index < 0 || index > size)
//            throw new IllegalArgumentException("Add failed. Illegal index.");
//
//        Node prev = dummyHead;
//        for(int i = 0 ; i < index ; i ++)
//            prev = prev.next;
//
//        prev.next = new Node(e, prev.next);
//        size ++;
//    }

    public void add(int index, E e){
        if(index< 0 || index > size){
            throw new IllegalArgumentException("Add failed. Illegal index.");
        }
        Node prev = dummyHead;
        for (int i=0; i< index ; i++){
            prev = prev.next;
        }

        prev.next = new Node(e, prev.next);
        size ++;

    }

    public E remove(int index){
        if(index< 0 || index >= size){
            System.out.println(" index: " + index);
            throw new IllegalArgumentException("remove failed. Illegal index.");
        }
        Node prev = dummyHead;
        for (int i=0; i< index ; i++){
            prev = prev.next;
        }
        Node delNode = prev.next;
        prev.next = delNode.next;
        E e = delNode.e;
        delNode = null;
        size --;
        return e;
    }
//
//    public E remove(int index){
//        if(index < 0 || index >= size)
//            throw new IllegalArgumentException("Remove failed. Index is illegal.");
//
//        Node prev = dummyHead;
//        for(int i = 0 ; i < index ; i ++)
//            prev = prev.next;
//
//        Node retNode = prev.next;
//        prev.next = retNode.next;
//        retNode.next = null;
//        size --;
//
//        return retNode.e;
//    }

    public E removeFirst(){
        return remove(0);
    }

    public E removeLast(){
        return remove(size-1);
    }




    public void addLast(E e){
        add(size, e);
    }

    public E get(int index){
        if(index< 0 || index >= size){
            throw new IllegalArgumentException("Add failed. Illegal index.");
        }
        Node cur = dummyHead.next;
        for (int i=0; i< index ; i++){
            cur = cur.next;
        }
        return cur.e;
    }


    public void set(int index, E e){
        if(index< 0 || index >= size){
            throw new IllegalArgumentException("Add failed. Illegal index.");
        }
        Node cur = dummyHead.next;
        for (int i=0; i< index ; i++){
            cur = cur.next;
        }
        cur.e = e;
    }

    public boolean contains(E e){

        Node cur = dummyHead.next;
        while(cur != null){
            if(cur.e == e){
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    @Override
    public String toString() {

        StringBuilder res = new StringBuilder();
        Node cur = dummyHead.next;
        while(cur != null){
            res.append(cur + "->");
            cur = cur.next;
        }
        res.append("NULL");
        return res.toString();
    }

    public static void main(String[] args) {

        LinkedList<Integer> linkedList = new LinkedList<>();
        for(int i = 0 ; i < 5 ; i ++){
            linkedList.addFirst(i);
            System.out.println(linkedList);
        }

        linkedList.add(2, 666);
        System.out.println(linkedList);

        linkedList.remove(2);
        System.out.println(linkedList);

        linkedList.removeFirst();
        System.out.println(linkedList);

        linkedList.removeLast();
        System.out.println(linkedList);
    }

}
